### ISSUE TITLE
is the first time user experience too mysterious?

### ISSUE THREAD

**Issue by LukeWinikates** [pivotal/LicenseFinder#235]
I asked my team to add license_finder to a gradle project and it was a rocky start. Turns out the problem was mostly #222. With the gradle discovery broken, the feedback from license_finder was misleadingly happy -- a few dots then exit `0`. If they were adding license_finder to the job without somebody around (me) who knows how the gem normally behaves, there wouldn't have been any way for them to know something was wrong.

I wonder if the default output should be more verbose, to make it more obvious when something is totally broken, and to set user expectations a bit better.

```
discovering package managers...
found gradle
found npm
discovering dependencies...
found 0 dependencies
```

**Comment by flavorjones** [comment]
@LukeWinikates this feels like a vote to fix #222, would you mind if I closed this?

**Comment by LukeWinikates** [comment]
It's not a duplicate of #222. I'm saying that the behavior of having effectively zero output if everything *seems* to go well makes the gem hard to use, and makes it hard to tell if something surprising has happened (like gradle not being discovered).

IOW, #222 is obviously a bug, but is it also a usability bug that #222 is effectively undiscoverable? Should the tool give more feedback about what it's doing, so that debugging is easier?

Package manager discovery is kind of magical but there's not a good way to sniff out whether the discovery is working or not.

**Comment by mainej** [comment]
Looks like it would be pretty easy to have
`LicenseFinder::CLI::Main#action_items` print a message when
`license_finder.acknowledged.empty?` It could also return a non-zero exit
code, though that may be incorrect because it really is possible to have no
dependencies. Maybe a PR would clarify the scope and utility of the change.

**Comment by flavorjones** [comment]
There's a `--debug` option, but it's incompletely and inconsistently implemented. In your case, @LukeWinikates, it should tell you that gradle isn't present.

I agree that output from LF is terrible, but unfortunately we have a very limited amount of engineering time that we can put towards overhauling UX. I'll leave this open.

**Comment by LukeWinikates** [comment]
I wasn't aware of the debug option, and I am happy to know it exists! Although Nuget is missing from the list it prints out... 🤔 

I'm interested in doing a PR. Honestly, I think my first suggestion would be to show the debug output, and possibly a count of the discovered dependencies per package manager. That would make the current output more verbose, but would make it more obvious what's happening under the hood.

I kind of want `license_finder` to print help text when invoked with no arguments, the way `git` or `cf` do. That would mean action_items wouldn't be the default command anymore though, and that would be a significant breaking change. So I suspect that the more verbose output for `action_items` would be more respectful to existing users.

**Comment by dkoper** [comment]
I ran into a similar issue: Ran `license-finder` on a Golang project but did not have golang itself installed.
LF returned with no output: means no action items, right? Wrong!
A note/warning that no dependencies were found would have helped :)

**Comment by LukeWinikates** [comment]
How about if `license_finder` prints a warning when it looks like a project is using a package manager, but the prerequisites aren't installed on the current machine? E.g. if we find `packages.json` but no `npm` executable, or in @dkoper 's situation where golang itself is not installed. I haven't looked at the code in a pretty long time, but this is probably a more modest change than my proposals from July. I find the code hard to follow, but I think that the `active?` check on each package manager conflates "does this project use this package manager?" with "does the environment have everything I need to interrogate the dependencies managed by this package manager?", and teasing apart that distinction would be valuable.

https://github.com/pivotal/LicenseFinder/blob/master/lib/license_finder/package_manager.rb#L19

I think we LF could also have a nonzero exit status for that case without needing to call that out as a breaking change.

I'll see if I can get some time to make a PR so that I'm not all talk and no action on this subject.
